kernel/copy.masm

```diff
@@ -0,0 +1,256 @@
+# assert_eqw(A: Word, B: Word)
+#
+# Asserts that the top two words in the stack are equal
+proc.assert_eqw.0 
+  movup.4 assert_eq movup.3 assert_eq movup.2 assert_eq assert_eq
+end
+
+# is_odd(el: Felt) -> Bool
+#
+# True if the element is odd (3 cycles)
+proc.is_odd.0
+  # input: [el, ...]
+  # output: [b, ...]
+
+  # drop the high bits
+  u32cast
+
+  # check if the odd bit is set
+  push.1 u32checked_and
+end
+
+# move_from_tape_to_memory_even_words(
+#   A: Word,
+#   B: Word,
+#   C: Word,
+#   writePtr: Address,
+#   words: Felt,
+# ) -> [A', B', C', writePtr']
+#
+# This is an optimized version of the inner loop.
+#
+# Given an initialized RPO state represented by [A, B, C], moves an even amount
+# `words` from the advice tape sequentially memory starting at address
+# `writePtr`. 
+#
+# Note:
+# - words must be even and greater than zero
+# - The words A,B,C compose the state of the RPO, these words must have been
+# properly initialized, including setting Capacity to [0,0,0,1] if the complete
+# copy is for an odd number of words
+#
+# Cycles: 7 + 8*words
+proc.move_from_tape_to_memory_even_words.0
+  # input: [A, B, C, writePtr, words, ...]
+  # output: [A, B, C, writePtr, ...]
+
+  # LOOP: [A, B, C, writePtr, words, ...]
+  # while(words>0)
+  dup.13 neq.0 while.true
+    # stack: [A', B', C', writePtr', words, ...]
+    adv_pipe
+
+    # move `words` counter up
+    # stack: [[writePtr', words, x, x], A', B', C', ...]
+    swapw.3
+
+    # stack: [[words, writePtr', x, x], A', B', C', ...]
+    movup.1
+
+    # optimization: `words` is required to be even and greater-than zero for
+    # this procedure to be used. Here it is possible to do a division instead
+    # of a range check, so the test is a single cycle
+    # stack: [[words/2, writePtr', x, x], A', B', C', ...]
+    div.2
+
+    # move `words` counter down
+    # LOOP: [A', B', C', writePtr', words/2, ...]
+    movdn.1 swapw.3
+
+    # while(words>0)
+    dup.13 neq.0
+  end
+
+  # stack: [A', B', C', writePtr', ...]
+  movup.13 drop
+end
+
+# move_from_tape_to_memory(words: Felt, writePtr: Address) -> [RpoHash, writePtr']
+#
+# Moves the `words` words from the advice tape to memory, starting at
+# `writePtr` address, the hash of the copied data is left on the stack for
+# commitment checking.
+proc.move_from_tape_to_memory.0
```

Comment:
Procedures intended for external consumption should be marked with `export` rather than `proc` keywords. So, for example, this one could be:
```
export.move_from_tape_to_memory
```